nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,186 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/keepout_filter.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: costmap_filter_info_sub_(nullptr), map_filter_sub_(nullptr), map_filter_(nullptr)
+{
+}
+
+void KeepoutFilter::initializeFilter(
+  const std::string costmap_filter_info_topic)
+{
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Setting a subscriber to %s costmap filter info",
+    costmap_filter_info_topic.c_str());
+  costmap_filter_info_sub_ = node_->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    costmap_filter_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::costmapFilterInfoCallback, this, std::placeholders::_1));
+
+  curr_time_ = node_->now();
+  prev_time_ = curr_time_;
+}
+
+void KeepoutFilter::costmapFilterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  // Resetting previous subscriber each time when new costmap filter information arrives
+  if (map_filter_sub_) {
+    RCLCPP_WARN(node_->get_logger(), "New map filter info arrived. Replacing old one.");
+    map_filter_sub_.reset();
+  }
+  map_filter_.reset();
+
+  // Setting new map filter subscriber
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Setting a subscriber to %s map filter",
+    msg->map_filter_topic.c_str());
+  map_filter_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    msg->map_filter_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::mapFilterCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::mapFilterCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  if (map_filter_) {
+    RCLCPP_WARN(node_->get_logger(), "New map filter arrived. Replacing old one.");
+    map_filter_.reset();
+  }
+  map_filter_ = std::make_unique<nav2_util::OccupancyGrid>(*msg);
```

Comment:
Why not just use the shared pointer the callback is giving you to work with? You can just store this map as a shared_ptr and not convert

Reply:
This was switched to a new constructor for Costmap2D with `OccupancyGrid` type: https://github.com/ros-planning/navigation2/commit/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-f67a16397bb5979b7cbf4729ea53a451R103